### PR TITLE
Update savings on staffapp

### PR DESCRIPTION
**Pull request by colinbruce**
Implement an improved savings flow for staff processing paper applications
===
[This ticket](https://www.pivotaltracker.com/story/show/117992675) improves the on screen experience to better match the paper form.

Still to do
===
- [x] Add tests for the WIP commit
- [x] Write a proper PR description
- [x] Refactor the pass/fail service
- [x] Improve the JS handling on the view
- [x] Check validation message changes
- [x] Mark as ready for review
- [ ] Merge in all subsequent PR's

### CODE REVIEW

File: app/models/forms/application/savings_investment.rb [hunk 1]
```diff
@@ -6,17 +6,18 @@ class SavingsInvestment < ::FormObject
 
       def self.permitted_attributes
         {
-          threshold_exceeded: Boolean,
-          partner_over_61: Boolean,
-          high_threshold_exceeded: Boolean
+          min_threshold_exceeded: Boolean,
+          over_61: Boolean,
+          max_threshold_exceeded: Boolean,
+          amount: Decimal
```
Comment (novotnyjakub) [inline app/models/forms/application/savings_investment.rb]: In the public app I assumed we're not dealing with pence, only pounds, so I used `Integer`. I'll try to find out what's correct.

File: app/models/saving.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+class Saving < ActiveRecord::Base
+  belongs_to :application, required: true, inverse_of: :saving
+  belongs_to :jurisdiction
```
Comment (novotnyjakub) [inline app/models/saving.rb]: I don't think this belongs here :)

File: app/services/partner_age_check.rb [hunk 3]
```diff
@@ -1,34 +0,0 @@
-# TODO: Make a validator out of this service
-class PartnerAgeCheck
```
Comment (novotnyjakub) [inline app/services/partner_age_check.rb]: 👍

File: app/services/savings_pass_fail_service.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
+class SavingsPassFailService
+
+  def initialize(saving)
+    @saving = saving
+  end
+
+  def calculate!
+    @saving.fee_threshold = FeeThreshold.new(@saving.application.fee).band
+    @saving.passed = calculate_pass_fail
+    unless @saving.passed
```
Comment (novotnyjakub) [inline app/services/savings_pass_fail_service.rb]: I would maybe abstract this to a method, to make the `calculate!` more readable. Something like `set_application_outcome_when_failed` or so.

File: app/services/savings_pass_fail_service.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
+class SavingsPassFailService
```
Comment (novotnyjakub) [inline app/services/savings_pass_fail_service.rb]: I don't see anywhere the min and max thresholds being stored on the `Saving` object. It's done in the migration for the existing applications, but not for the new ones. Unless I missed something?

File: spec/models/views/overview/savings_and_investments_spec.rb [hunk 6]
```diff
@@ -3,62 +3,62 @@
...
-
-    context 'when user selected "no" to threshold_exceeded' do
-      let(:application) { build_stubbed :married_applicant_under_61, threshold_exceeded: false }
-
-      it { is_expected.to eq nil }
-    end
+    it { is_expected.to eql '£3500' }
   end
+  #
+  # describe '#combined_savings_valid?' do
```
Comment (novotnyjakub) [inline spec/models/views/overview/savings_and_investments_spec.rb]: Let's remove this commented stuff, if it's not needed.

File: app/services/savings_pass_fail_service.rb (hunk 5 above)
Comment (CeeBeeUK) [inline app/services/savings_pass_fail_service.rb]: They are saved in the form object, before this service is called

Comment (colinbruce) [inline app/services/savings_pass_fail_service.rb]: They are saved in the form object, before this service is called

Comment (novotnyjakub) [inline app/services/savings_pass_fail_service.rb]: They don't seem to be: 

```ruby
       def fields_to_update
          {
           min_threshold_exceeded: min_threshold_exceeded,
           over_61: over_61,
           max_threshold_exceeded: max_threshold_exceeded,
           amount: amount
          }
        end
```

Comment (colinbruce) [inline app/services/savings_pass_fail_service.rb]: You, sir, are correct!  Two many similarly named fields!   Yes, I was going to add a setting for min and max values and import them... 
On my todo list now! :)

Comment (novotnyjakub) [inline app/services/savings_pass_fail_service.rb]: Yeah, I missed it originally as well, when scanning through it the first time.

**Comment by novotnyjakub** [comment]
@colinbruce one important thing I realised is that we can't merge this until we adjust the public submission endpoint.. Let's decide how to do that on Monday..

**Comment by axemonkey** [comment]
JS fixed @colinbruce @novotnyjakub 👍

**Comment by joshlow** [comment]
@colinbruce - copy suggestions for error messages

Please answer the savings question

Please enter the amount of savings and investments

Please confirm if the partner is 61 or over

Please answer the savings question

**Pull request merged**